sql-plugin/src/main/scala/com/nvidia/spark/rapids/BatchWithPartitionData.scala

```diff
@@ -0,0 +1,424 @@
+   *  - resultBatches: Array that stores the resulting batches after splitting.
+   *
+   * Algorithm:
+   *  - Initialize `sizeOfBatch` and `resultBatches`.
+   *  - Iterate through `partRows`:
+   *    - Get rowsInPartition - This can either be rows from new partition or
+   *      rows remaining to be processed if there was a split.
+   *    - For each column, calculate the size of the partition.
+   *    - Check if adding any partition to the current batch would exceed batch size limit.
+   *    - If it does, split the partition:
+   *      - Calculate max num of rows we can add to current batch.
+   *      - Append entry (InternalRow, max rows that fit) to the current batch.
+   *      - Append current batch to the result batches.
+   *      - Reset variables.
+   *    - If there was no split,
+   *      - Append entry (InternalRow, rowsInPartition) to the current batch.
+   *      - Update sizeOfBatch with size of partition for each column.
+   *      - This implies all remaining rows can be added in current batch without exceeding limit.
+   *
+   * Example:
+   * {{{
+   * Input:
+   *    partition rows:   [10, 40, 70, 10, 11]
+   *    partition values: [ [abc, ab], [bc, ab], [abc, bc], [aa, cc], [ade, fd] ]
+   *    limit:  300 bytes
+   *
+   * Result:
+   * [
+   *  [ (10, [abc, ab]), (40, [bc, ab]), (63, [abc, bc]) ],
+   *  [ (7, [abc, bc]), (10, [aa, cc]), (11, [ade, fd]) ]
+   * ]
+   * }}}
+   *
+   * @return An array of batches, containing (row counts, partition values) pairs,
+   *         such that each batch's size is less than batch size limit.
+   */
+  private def splitPartitionDataIntoGroups(
+      partitionRowData: Array[PartitionRowData],
+      partSchema: StructType,
+      maxSizeOfBatch: Long): Array[Array[PartitionRowData]] = {
+    val allowedMaxSizeOfBatch = Math.min(CUDF_COLUMN_SIZE_LIMIT, maxSizeOfBatch)
+    val resultBatches = ArrayBuffer[Array[PartitionRowData]]()
+    val currentBatch = ArrayBuffer[PartitionRowData]()
+    val sizeOfBatch = Array.fill(partSchema.length)(0L)
+    var partIndex = 0
+    var splitOccurred = false
+    var rowsInPartition = 0
+    while (partIndex < partitionRowData.length) {
+      // If there was no split, get next partition else process remaining rows in current partition.
+      if (!splitOccurred) {
+        rowsInPartition = partitionRowData(partIndex).rowNum
+      }
+      val valuesInPartition = partitionRowData(partIndex).rowValue
+      val sizeOfPartition = partSchema.zipWithIndex.map { case (field, colIndex) =>
+        calculatePartitionSizes(rowsInPartition, valuesInPartition, field, colIndex)
+      }
+      splitOccurred = sizeOfBatch.zip(sizeOfPartition).exists {
+        case (batchSize, partitionSize) => batchSize + partitionSize > allowedMaxSizeOfBatch
+      }
+      if (splitOccurred) {
+        // Calculate max rows that can fit.
+        val maxRows = calculateMaxRows(sizeOfBatch, valuesInPartition, allowedMaxSizeOfBatch)
```

Comment:
It's a little odd to calculate the sizes and then, sorta redundantly, calculate the maximum rows.  Seems like this could be made more efficient by having a single utility method, something like calculateMaxRows, that computes the maximum number of rows that can can be used across any column.  We always compute the max rows for the current partition.  A split occurs when the resulting maxRows is less than the number of rows we would expect when taking the entire partition, and then we already know where to perform the split as well.